### ISSUE TITLE
Made data transfer more transparent

### ISSUE THREAD

**Issue by Mannshoch** [andyholmes/gnome-shell-extension-gsconnect#351]
Please add status icon if Data-transfer is complete.

How I would Imagine this:
1. I do a Data-transfer e.g. press CTRL+C 
1. ![grafik](https://user-images.githubusercontent.com/7795891/48851851-5457d400-edac-11e8-9a55-ba60474aabc7.png) appear and indicate that data is sending
1. ![grafik](https://user-images.githubusercontent.com/7795891/48851886-6f2a4880-edac-11e8-9b0a-e98fd964577e.png) appear and tell me the transfer was successful.
1. ![grafik](https://user-images.githubusercontent.com/7795891/48852038-c3352d00-edac-11e8-9b1c-4aef1955d1c9.png) appear if transfer failed.
1. ![grafik](https://user-images.githubusercontent.com/7795891/48852356-77cf4e80-edad-11e8-80b9-322cb950c0a3.png) If I hover over a clipboard entry in gsconnect I get a tool-tip with the last sent or received clipboard entry. If I click on it the last copied entry it get again into the clipboard and get shared.

Background:
In earlier version of gsconnect and KDEconnect the clipboard sharing worked amazing I linked two computer and two mobile-phone together with KDE connect and gsconnect. And where ever I copied a text it was nearly immediately available on all other devices. The Idea I write down raised this times already. 
Because of personal matters I didn't use it the last three month. I did updates on all devices and now it hardly work - sometimes. I'm not sure where the Problem is and why it happened. (the connection work fluently and e.g. the data-transfer also work (if the file not bigger then around 10MB).
If needed I wil write also a bug-report for this

**Comment by andyholmes** [comment]
Sorry, I'm not sure I understand the problem exactly. Are you copying very large pieces of text using the Clipboard plugin and it is taking a long time?

Clipboard "upload"/"download" is done using simple packets like this:

```js
{
  "id": 1542816155790,
  "type": "kdeconnect.clipboard",
  "body": {
    "content": "text being copied"
  }
}
```
When "uploading" packets are sent immediately, but there is no response or confirmation they have been received. When "downloading" packets are read entirely until finished and there is no way to know a packet in transfer is a `kdeconnect.clipboard` packet until it is finished.

So there is never any progress update or way to know if the packet was sent or received successfully. Does that make sense?

**Comment by Mannshoch** [comment]
Hallo Andy
Thanks for the response. Yes it make sense. If you do not get any feedback, that mad it a bit difficult to get feedback about a successful transfer.

I do not think that I transfer long text. Most of Time at moment, I copy some Links or small Text until around 60character.

**Comment by ferdnyc** [comment]
Small bits of text like that should be transferred in a single packet, they _should_ still be as instantaneous as you experienced in the previous version. If they're not, then something must be wrong with the connection (meaning, either GSConnect or KDE Connect is doing something wrong that's delaying the transfers.)

It might be helpful if you could turn on the Debug mode [see this wiki article for more info](https://github.com/andyholmes/gnome-shell-extension-gsconnect/wiki/Debugging), then perform a couple of clipboard operations with data that's not particularly sensitive, and see what's displayed.

When I do that, I almost instantaneously (within milliseconds) see a `[core.js:send:450]` message containing the contents of the clipboard, as it's being sent to my phone:
```
DEBUG: [core.js:send:450]: Galaxy S6: {
  "id": 1542978492572,
  "type": "kdeconnect.clipboard",
  "body": {
    "content": "307.5G"
  }
}
```
Longer clipboard strings show up just as quickly:
```
DEBUG: [core.js:send:450]: Galaxy S6: {
  "id": 1542978695399,
  "type": "kdeconnect.clipboard",
  "body": {
    "content": "Error loading '/usr/share/gimp/2.0/tool-presets/GPS-Set1/FX-Radial-Softlight.gtp': Error while parsing 'stream' in line 5: unable to determine type of 'GimpBlendOptions'"
  }
}
```

If there's a delay in those messages being generated, then the log may give us some idea of why. If there _isn't_ a delay, then it's possible things are being held up on the KDE Connect end of the connection for some reason.

(Oh, P.S — don't forget to switch off Debugging once you're finished capturing the log output!)

**Comment by andyholmes** [comment]
> I do not think that I transfer long text. Most of Time at moment, I copy some Links or small Text until around 60character.

@Mannshoch Hmm, that should take no time at all. I would guess this might be happening when your phone is in "doze" (sleep) mode?

> If there's a delay in those messages being generated, then the log may give us some idea of why.

@ferdnyc maybe at some point we could implement a better debug infrastructure. The current one is pretty simple, for the reason that I wanted it to be easily toggled, but with very little performance impact both when on and off. It's actually better than the GLib functions (I think) because turning it off swaps in an empty function (`() => {}`) which is about as close to a no-op as I could get.

**Comment by ferdnyc** [comment]
> maybe at some point we could implement a better debug infrastructure. The current one is pretty simple, for the reason that I wanted it to be easily toggled, but with very little performance impact both when on and off. It's actually better than the GLib functions (I think) because turning it off swaps in an empty function (`() => {}`) which is about as close to a no-op as I could get.

It might be nice to have a few different classes of debug message, and a bitmask-or-equivalent to selectively enable/disable parts, so that you could turn on just packet dumps, or plugin debugging, or what have you. Keep things a little more focused when you want to focus on debugging a particular area or aspect of the code. ...At the same time, it's easy to overcomplicate that sort of thing, and build in tons of flexibility that never ends up getting used for any practical purpose.

Oh, FYI — speaking of `gnome-terminal`, [I just filed this](https://gitlab.gnome.org/GNOME/gnome-terminal/issues/58) — oh, and looks like it's already been answered. Despite the messages `gnome-termnal` spits out, sounds like `--command` isn't going away after all, at least not right away. So, that's cool.

**Comment by ferdnyc** [comment]
> I realize that the https://extensions.gnome.org website is very stingy with its data. I do not see the actual Versions number of gsconnect and do not see their publish date or a changelog.

They actually did finally add a method of at least _determining_ what version(s) they have available, in the form of the direct Download option — if you select your (GNOME) "Shell Version..." in the first dropbox, the second will show you all of the extension versions they've reviewed and approved for that release of GNOME Shell. (Oldest-on-top, which is very unintuitive since the Shell Version dropbox is newest-on-top!) So, right now, users running 3.30 can download GSConnect v13, v15, v16, or v17. Anyone being offered an update should be getting v17 (if they're running GNOME Shell 3.28 or later).

But it's an incredibly roundabout way of getting at that basic information, and you're absolutely right that they really don't **volunteer** much useful data at all. Nor do they offer the ability for extension authors to publish changelogs directly through e.g.o, when submitting extension updates.

I think the idea was to have Gnome Software handle extension updates with full metadata (including changelogs), but since Gnome Software currently can't even detect the _version number_ of extensions I've installed from e.g.o, that doesn't seem likely to happen anytime soon.

**Comment by andyholmes** [comment]
Since it's been over two weeks since the last activity, I'm going to close this bug. I believe this was primarily a misunderstanding about how the clipboard plugin operates and/or a network related issue.

Please re-open if the issue persists with a current version of GSConnect.
